app/routes/start-application.js

```diff
@@ -23,22 +29,38 @@ module.exports = [{
   path: '/start-application',
   options: {
     validate: {
-      payload: schema
+      payload: Joi.object()
+        .concat(sbiSchema)
+        .concat(agreementNumberSchema)
     },
     handler: async (request, h) => {
-      let { agreement, data, callerId, crn } = await cache.get(request)
+      let agreement
+      const { data, callerId, crn } = await cache.get(request)
+      const sbi = parseInt(request.payload.sbi)
 
-      // if no existing agreement or SBI no longer matches cached agreement then start new agreement
-      if (!agreement || agreement?.organisation?.sbi !== request.payload.sbi) {
-        await cache.reset(request)
+      await cache.reset(request)
+      if (!request.payload.agreementNumber) {
         agreement = create()
-        // Need to include caller Id as part of agreement to support downstream services
-        // Expect to remove once Defra Identity integrated
-        agreement.callerId = callerId
-        agreement.crn = crn
-      }
+      } else {
+        const incompleteApplications = await getBySbi(sbi)
```

Comment:
Instead of using getBySbi and having to check the array you can use getByAgreementNumber. Also you should be able to cut down on building up the agreement object:

```
      const sbi = parseInt(request.payload.sbi)
      const agreementNumber = request.payload.agreementNumber

      await cache.reset(request)
      if (!agreementNumber) {
        agreement = create()
        // Need to include caller Id as part of agreement to support downstream services
        // Expect to remove once Defra Identity integrated
        agreement.callerId = callerId
        agreement.crn = crn
        const selectedOrganisation = data.eligibleOrganisations.find(x => x.sbi === sbi)

        if (selectedOrganisation) {
          agreement.organisation = selectedOrganisation
        }
      } else {
        const incompleteApplications = await getByAgreementNumber(agreementNumber, sbi)

        if (!incompleteApplications) {
          return h.view('404')
        }

        agreement = incompleteApplications.agreementData
      }

      if (agreement.organisation) {
        await cache.update(request, { agreement })
        return h.redirect('/task-list')
      }

      return h.redirect('/eligible-organisation')
    }
```

Reply:
I think the reason why we used getBySbi is because we need the list of agreementNumbers so that we can display the table. If we used getByAgreementNumber we would lose that information.

Also thanks for this:
`agreement = incompleteApplications.agreementData`

we didn't think was an easier way!

Reply:
Yeah, that is very true, sorry. I thought you were also suggesting it in the GET route. We used the `getByAgreementNumber` for POST in the end so I think it aligns with what you were referring to.